### PR TITLE
[Refactor] Refactor parquet reader create logic

### PR DESCRIPTION
**Pull request by Smith-Cruise**
## Why I'm doing:
Using ColumnReader to handle mapping relationships(instead of MetaHelper), so we can handle complex type's statistics more easily.

## What I'm doing:

* GroupReader/ColumnReader have two phase: `init()` -> `prepare()`. In `init()`, we will create ColumnReader and bind ParquetField/ColumnMetaData to each ColumnReader. In `prepare()` we create StoredColumnReader in ColumnReader. We have to make sure init() is lightweight because we may filter the entire GroupReader after evaluating with statistics.
* Each ColumnReader will provide `get_chunk_metadata()` and `get_column_parquet_field()` method. For complex type ColumnReader, `get_chunk_metadata()` will always return nullptr, because it didn't have ColumnMetaData.
* Move ColumnReader create logic to ColumnReaderFactory.
* Move FileMetaData parse logic into FileMetaDataParser.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/formats/parquet/group_reader.cpp [hunk 1]
```diff
@@ -286,21 +308,23 @@ Status GroupReader::_init_column_readers() {
...
     }
     return Status::OK();
 }
 
-Status GroupReader::_create_column_reader(const GroupReaderParam::Column& column) {
+StatusOr<ColumnReaderPtr> GroupReader::_create_column_reader(const GroupReaderParam::Column& column) {
     std::unique_ptr<ColumnReader> column_reader = nullptr;
     const auto* schema_node = _param.file_metadata->schema().get_stored_column_by_field_idx(column.idx_in_parquet);
     {
-        SCOPED_RAW_TIMER(&_param.stats->column_reader_init_ns);
```
Comment (Smith-Cruise) [inline be/src/formats/parquet/group_reader.cpp:298]: TIMER move to _create_column_readers()

File: be/src/formats/parquet/group_reader.cpp [hunk 2]
```diff
@@ -259,23 +296,8 @@ StatusOr<size_t> GroupReader::_read_range_round_by_round(const Range<uint64_t>&
     return hit_count;
 }
 
-void GroupReader::close() {
```
Comment (Smith-Cruise) [inline be/src/formats/parquet/group_reader.cpp:262]: move to deconstructor

File: be/src/formats/parquet/column_reader_factory.cpp [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+#include "formats/parquet/column_reader_factory.h"
+
+#include "formats/parquet/complex_column_reader.h"
+#include "formats/parquet/scalar_column_reader.h"
+#include "formats/parquet/schema.h"
+#include "formats/utils.h"
+
+namespace starrocks::parquet {
+
+StatusOr<ColumnReaderPtr> ColumnReaderFactory::create(const ColumnReaderOptions& opts, const ParquetField* field,
```
Comment (Smith-Cruise) [inline be/src/formats/parquet/column_reader_factory.cpp:24]: code is copy from `column_reader.cpp`

File: be/src/formats/parquet/group_reader.cpp [hunk 4]
```diff
@@ -47,12 +48,27 @@ GroupReader::GroupReader(GroupReaderParam& param, int row_group_number, const st
     _row_group_metadata = &_param.file_metadata->t_metadata().row_groups[row_group_number];
 }
 
-Status GroupReader::init() {
-    // the calling order matters, do not change unless you know why.
-    RETURN_IF_ERROR(_init_column_readers());
-    _process_columns_and_conjunct_ctxs();
-    _range = SparseRange<uint64_t>(_row_group_first_row, _row_group_first_row + _row_group_metadata->num_rows);
+GroupReader::~GroupReader() {
+    if (_param.sb_stream) {
```
Comment (Smith-Cruise) [inline be/src/formats/parquet/group_reader.cpp:52]: copy code from `close()`

File: be/src/formats/parquet/group_reader.cpp [hunk 5]
```diff
@@ -73,6 +89,9 @@ Status GroupReader::_deal_with_pageindex() {
 }
 
 Status GroupReader::prepare() {
+    RETURN_IF_ERROR(_prepare_column_readers());
```
Comment (Smith-Cruise) [inline be/src/formats/parquet/group_reader.cpp:92]: Move from `GroupReader()::init()`

File: be/src/formats/parquet/file_reader.cpp [hunk 6]
```diff
@@ -482,11 +280,10 @@ bool FileReader::_filter_group_with_more_filter(const tparquet::RowGroup& row_gr
                     LOG(WARNING) << "couldn't find slot id " << kv.first << " in tuple desc";
                     continue;
                 }
-                std::unordered_map<std::string, size_t> column_name_2_pos_in_meta{};
-                std::vector<SlotDescriptor*> slot_v{slot};
-                _meta_helper->build_column_name_2_pos_in_meta(column_name_2_pos_in_meta, slot_v);
-                const tparquet::ColumnMetaData* column_meta =
-                        _meta_helper->get_column_meta(column_name_2_pos_in_meta, row_group, slot->col_name());
+                const tparquet::ColumnMetaData* column_meta = nullptr;
+                if (group_reader->get_chunk_metadata(slot->id())->__isset.meta_data) {
```
Comment (Youngwb) [inline be/src/formats/parquet/file_reader.cpp]: why need this check? when ColumnChunk not set meta_data?

**Review by Youngwb — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Review by zombee0 — APPROVED** [review]

**Comment by trueeyu** [comment]
https://github.com/Mergifyio backport branch-3.3

**Comment by trueeyu** [comment]
https://github.com/Mergifyio backport branch-3.3

**Pull request merged**